### PR TITLE
Add fourier features for long seasonality time series for arima method

### PR DESCRIPTION
**Pull request by leica2023**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R [hunk 1]
```diff
@@ -3,15 +3,22 @@ loadNamespace("forecast")
 handleForecast <- function(model, params) {
     outputs = list()
     output_types = params$output_types
+
+    if (("xreg_future" %in% names(params)) & ("xreg" %in% names(model))) {
+    regressor_future = params$xreg_future
+    } else {
+    regressor_future = NULL
+    }
```
Comment (lostella) [inline src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R:11]:
```suggestion
    if (("xreg_future" %in% names(params)) & ("xreg" %in% names(model))) {
        regressor_future = params$xreg_future
    } else {
        regressor_future = NULL
    }
```

File: src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R [hunk 2]
```diff
@@ -52,12 +59,17 @@ handlePointForecast <- function(forecasts, params) {
 
 
 arima <- function(ts, params) {
+    if ('xreg' %in% names(params)) {
+    model <- forecast::auto.arima(ts, trace=TRUE, xreg = params$xreg, seasonal = params$seasonal)
+    } else {
     model <- forecast::auto.arima(ts, trace=TRUE)
+    }
```
Comment (lostella) [inline src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R:66]:
```suggestion
    if ('xreg' %in% names(params)) {
        model <- forecast::auto.arima(ts, trace=TRUE, xreg = params$xreg, seasonal = params$seasonal)
    } else {
        model <- forecast::auto.arima(ts, trace=TRUE)
    }
```

File: src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R [hunk 3]
```diff
@@ -3,15 +3,22 @@ loadNamespace("forecast")
...
     if ("samples" %in% output_types) {
-        outputs$samples <- lapply(1:params$num_samples, function(n) { simulate(model, params$prediction_length) } )
+        outputs$samples <- lapply(1:params$num_samples, function(n) { simulate(model, params$prediction_length, xreg = regressor_future)})
     }
     if("quantiles" %in% output_types) {
-        f_matrix <- forecast::forecast(model, h=params$prediction_length, level=unlist(params$levels))$upper
-        outputs$quantiles <- split(f_matrix, col(f_matrix))
+        f_matrix <- forecast::forecast(model, h=params$prediction_length, xreg = regressor_future, level=unlist(params$intervals))
+        outputs$upper_quantiles <- split(f_matrix$upper, col(f_matrix$upper))
+        outputs$lower_quantiles <- split(f_matrix$lower, col(f_matrix$lower))
```
Comment (lostella) [inline src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R:18]: Is the `xreg = regressor_future` not enough here?

File: src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R [hunk 4]
```diff
@@ -52,12 +59,17 @@ handlePointForecast <- function(forecasts, params) {
...
+    } else {
     model <- forecast::auto.arima(ts, trace=TRUE)
+    }
     handleForecast(model, params)
 }
 
 ets <- function(ts, params) {
-    model <- forecast::ets(ts, additive.only=TRUE)
+
+    model <- forecast::ets(ts, additive.only=TRUE)
```
Comment (lostella) [inline src/gluonts/ext/r_forecast/R/univariate_forecast_methods.R:72]: Change not needed?

File: src/gluonts/time_feature/_base.py [hunk 5]
```diff
@@ -163,6 +163,31 @@ def week_of_year_index(index: pd.PeriodIndex) -> np.ndarray:
         week = index.week
     return week.astype(float).values - 1
 
+class fourier_time_features:
```
Comment (lostella) [inline src/gluonts/time_feature/_base.py:166]: I think you can keep this definition local to the `gluonts.ext.r_forecast` package for now. Also, we use the CamelCase convention for class names

File: src/gluonts/ext/r_forecast/_predictor.py [hunk 6]
```diff
@@ -317,15 +316,20 @@ def predict(
 
         """
         self._warning_message()
-
+        print(f"params is {self.params}")
```
Comment (leica2023) [inline src/gluonts/ext/r_forecast/_predictor.py:319]: remove this

File: src/gluonts/ext/r_forecast/_predictor.py [hunk 7]
```diff
@@ -317,15 +316,20 @@ def predict(
...
         for data in dataset:
+
+            if self.method_name == 'arima':
+                target_len = data['target'].shape[0]
+                freq = data['start'].freq
+                self.params = fourier_time_features(target_len, freq, self.prediction_length)(self.params)
```
Comment (lostella) [inline src/gluonts/ext/r_forecast/_predictor.py:325]: maybe do
```suggestion
               params = fourier_time_features(target_len, freq, self.prediction_length)(self.params)
```
here, so as not to mutate the predictor object

File: src/gluonts/ext/r_forecast/_predictor.py [hunk 8]
```diff
@@ -317,15 +316,20 @@ def predict(
 
         """
         self._warning_message()
-
+        print(f"params is {self.params}")
         for data in dataset:
+
+            if self.method_name == 'arima':
```
Comment (leica2023) [inline src/gluonts/ext/r_forecast/_predictor.py:322]: if self.use_fourier_features == True:

File: src/gluonts/time_feature/_base.py [hunk 9]
```diff
@@ -163,6 +163,31 @@ def week_of_year_index(index: pd.PeriodIndex) -> np.ndarray:
...
+    fourier_frequency_low_periods = 4
+    fourier_ratio_threshold_low_periods = 18
+    fourier_frequency_high_periods = 52
+    fourier_ratio_threshold_high_periods = 2
+    fourier_order = 4
+    def __init__(self, target_len, freq, prediction_length):
+        self.target_len = target_len
+        self.freq = freq
+        self.prediction_length = prediction_length
+    def __call__(self, params):
```
Comment (lostella) [inline src/gluonts/time_feature/_base.py:177]: Could this object just be a function then?

File: src/gluonts/ext/r_forecast/_predictor.py (hunk 8 above)
Comment (leica2023) [reply src/gluonts/ext/r_forecast/_predictor.py:322]: use_fourier_features should be an argument in use_fourier_features

**Review by lostella — COMMENTED** [review]
As per offline discussion: the change as it currently is in the PR is too much ad-hoc. Better ways to include this probably are 
- either add support for `feat_dynamic_real` to the `RForecastPredictor` class, where this is passed on to the model as `xreg` and `xreg_future`. This way, one can add whatever set of features from the outside, not just limited to seasonal features
- or, implement a new R method in `univariate_forecast_method.R`, where the desired hard-coded logic for computing `xreg` and `xreg_future` is implemented in R directly, and those are then passed on to `forecast::auto.arima` and `handleForecast` respectively.

**Pull request closed without merging**